### PR TITLE
snmp_listener: add missing tags config to snmp listener

### PR DESCRIPTION
**Pull request by AlexandreYang**
### What does this PR do?

snmp: add extra tags to snmp listener config

### Motivation

user request to have subnet specific tags to be passed as instance tags

### Why this is a bug for 7.27

snmp_listener (Agent SNMP Autodiscovery) has been already released prior 7.27 and SNMP corecheck is being released 7.27.

Currently users are using Python SNMP Autodiscovery that has network/subnet tags support but has important performance limitations.
Some users like Wayfair need SNMP Autodiscovery (Python or Agent), so switching to SNMP corecheck also means using Agent SNMP Autodiscovery (snmp_listener).
Note: SNMP corecheck itself does not provide itself AD support.

But Agent SNMP Autodiscovery + SNMP corecheck lacks network/subnet tags as mentioned in the context above, and does not provide feature parity with Python AD that can be considered as a bug.

### Additional Notes

Example of snmp listener config:

```yaml
snmp_listener:
  configs:
    - network: 192.168.1.16/29
      version: 2
      port: 1161
      community: cisco_icm
      tags:
      - "tag1:val1"
      - "tag2:val2"
```

with template looking like this:

```yaml
ad_identifiers:
  - snmp
init_config:
instances:
  -
    ## @param ip_address - string - optional
    ## The IP address of the device to monitor.
    #
    ip_address: "%%host%%"
    
    […]

    ## @param extra_tags - string - optional
    ## Comma separated tags to attach to every metric, event and service check emitted by this integration.
    ## Example:
    ##  extra_tags: "tag1:val1,tag2:val2"
    #
    extra_tags: "%%extra_tags%%"
```

This template will expand to instance config like this:
```yaml
    ip_address: 192.168.1.17
    ...
    extra_tags: "tag1:val1,tag2:val2"
```

Then the integration will split comma separated `extra_tags` and use the tags on all metrics and service checks (events, if any).

### Describe your test plan

Test using a `snmp_listener.configs` config with `extra_tags` like this:

```yaml
snmp_listener:
  configs:
    - network: 192.168.1.16/29
      version: 2
      port: 1161
      community: cisco_icm
      tags:
      - "tag1:val1"
      - "tag2:val2"
```

Check if metrics reported by snmp integration contain the tags from snmp_listener config.

#### 1/ Test with single agent autodiscovery using snmp_listener: Setup guide: https://datadoghq.dev/integrations-core/architecture/snmp/#standalone-agent

Can be tested using:

a/

Add tags (PR feature) to this config locally: https://github.com/DataDog/integrations-core/blob/41f3e16b7483846a453e3bbef5348355fcbf8db6/snmp/tests/conftest.py#L95-L100

b/
```
ddev env start snmp py38-agent_ad -a  <RC docker image tag>
```

c/

```
ddev env check snmp py38-agent_ad
# or from insider container using `agent check snmp` and check if tags you added in 1/ are there.
```

#### 2/ Test with DCA autodiscovery using snmp_listener. Setup guide: https://datadoghq.dev/integrations-core/architecture/snmp/#cluster-agent-support

Prerequisite: local kubernetes (e.g. docker-desktop)

a/

Copy following yaml into a file `cluster-agent-values_snmp_listener.yaml`:

<details>
  <summary>Helm Chart values</summary>

```
datadog:
  ## @param apiKey - string - required
  ## Set this to your Datadog API key before the Agent runs.
  ## ref: https://app.datadoghq.com/account/settings#agent/kubernetes
  #
  apiKey: <DATADOG_API_KEY>

  ## @param clusterName - string - optional
  ## Set a unique cluster name to allow scoping hosts and Cluster Checks easily
  ## The name must be unique and must be dot-separated tokens where a token can be up to 40 characters with the following restrictions:
  ## * Lowercase letters, numbers, and hyphens only.
  ## * Must start with a letter.
  ## * Must end with a number or a letter.
  ## Compared to the rules of GKE, dots are allowed whereas they are not allowed on GKE:
  ## https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1beta1/projects.locations.clusters#Cluster.FIELDS.name
  #
  clusterName: snmp-perf-lab-cluster

  ## @param clusterChecks - object - required
  ## Enable the Cluster Checks feature on both the cluster-agents and the daemonset
  ## ref: https://docs.datadoghq.com/agent/autodiscovery/clusterchecks/
  ## Autodiscovery via Kube Service annotations is automatically enabled
  #
  clusterChecks:
    enabled: true

  ## @param tags  - list of key:value elements - optional
  ## List of tags to attach to every metric, event and service check collected by this Agent.
  ##
  ## Learn more about tagging: https://docs.datadoghq.com/tagging/
  #
  tags:
    - 'env:test-snmp-cluster-agent'

  logLevel: info

  ## Enable process agent and provide custom configs
  processAgent:
    # datadog.processAgent.enabled -- Set this to true to enable live process monitoring agent
    ## Note: /etc/passwd is automatically mounted to allow username resolution.
    ## ref: https://docs.datadoghq.com/graphing/infrastructure/process/#kubernetes-daemonset
    enabled: false

agents:
  # agents.enabled -- You should keep Datadog DaemonSet enabled!
  ## The exceptional case could be a situation when you need to run
  ## single Datadog pod per every namespace, but you do not need to
  ## re-create a DaemonSet for every non-default namespace install.
  ## Note: StatsD and DogStatsD work over UDP, so you may not
  ## get guaranteed delivery of the metrics in Datadog-per-namespace setup!
  #
  enabled: true

  ## Define the Datadog image to work with
  image:
    # agents.image.repository -- Datadog Agent image repository to use
    ## use "gcr.io/datadoghq/agent" for Datadog Agent 7
    ## use "gcr.io/datadoghq/dogstatsd" for Standalone Datadog Agent DogStatsD 7
    repository: datadog/agent-dev

    # agents.image.tag -- Define the Agent version to use
    ## Use 7-jmx to enable jmx fetch collection
    tag: alex-snmp-add-tags-to-snmp-listener-config-extra-tags-py3
    doNotCheckTag: true

  containers:
    agent:
      ## @param env - list - required
      ## Additional environment variables for the agent container.
      #
      env:
      - name: DD_KUBELET_TLS_VERIFY
        value: "false"
      - name: DD_CHECK_RUNNERS
        value: "16"

      ## @param livenessProbe - object - required
      ## Override the agent's liveness probe logic from the default:
      ## In case of issues with the probe, you can disable it with the
      ## following values, to allow easier investigating:
      #
      livenessProbe:
        httpGet:
          path: /live
          port: 5555
        initialDelaySeconds: 60
        periodSeconds: 60
        timeoutSeconds: 30
        successThreshold: 1
        failureThreshold: 6

      ## @param readinessProbe - object - required
      ## Override the agent's readiness probe logic from the default:
      ## In case of issues with the probe, you can disable it with the
      ## following values, to allow easier investigating:
      #
      readinessProbe:
        httpGet:
          path: /ready
          port: 5555
        initialDelaySeconds: 60
        periodSeconds: 60
        timeoutSeconds: 30
        successThreshold: 1
        failureThreshold: 6

## @param clusterAgent - object - required
## This is the Datadog Cluster Agent implementation that handles cluster-wide
## metrics more cleanly, separates concerns for better rbac, and implements
## the external metrics API so you can autoscale HPAs based on datadog metrics
## ref: https://docs.datadoghq.com/agent/kubernetes/cluster/
#
clusterAgent:
  ## @param enabled - boolean - required
  ## Set this to true to enable Datadog Cluster Agent
  #
  enabled: true

  env:
  - name: DD_LOG_LEVEL
    value: "debug"
  ## Define the Datadog Cluster-Agent image to work with
  image:
    # clusterAgent.image.name -- Cluster Agent image name to use (relative to `registry`)
    name: cluster-agent-dev

    # clusterAgent.image.tag -- Cluster Agent image tag to use
    tag: alex-snmp-add-tags-to-snmp-listener-config-extra-tags

    # clusterAgent.image.repository -- Override default registry + image.name for Cluster Agent
    repository: datadog/cluster-agent-dev

    # clusterAgent.image.pullPolicy -- Cluster Agent image pullPolicy
    pullPolicy: IfNotPresent

    # clusterAgent.image.pullSecrets -- Cluster Agent repository pullSecret (ex: specify docker registry credentials)
    ## See https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod
    pullSecrets: []
    #   - name: "<REG_SECRET>"


  # cluster-agent-values_snmp_listener.yaml

  # Enable the metricsProvider to be able to scale based on metrics in Datadog
  metricsProvider:
    # clusterAgent.metricsProvider.useDatadogMetrics -- Enable usage of DatadogMetric CRD to autoscale on arbitrary Datadog queries
    ## NOTE: You need to install the `DatadogMetric` CRD before
    useDatadogMetrics: false

  ## @param confd - list of objects - optional
  ## Provide additional cluster check configurations
  ## Each key will become a file in /conf.d
  ## ref: https://docs.datadoghq.com/agent/autodiscovery/
  #
  confd:
     # Static checks
     http_check.yaml: |-
       cluster_check: true
       instances:
         - name: 'Check Example Site1'
           url: http://example.net
     snmp.yaml: |-
      cluster_check: true
      ad_identifiers:
        - snmp
      init_config:
      instances:
        -
          ## @param ip_address - string - optional
          ## The IP address of the device to monitor.
          #
          ip_address: "%%host%%"

          ## @param port - integer - optional - default: 161
          ## Default SNMP port.
          #
          port: "%%port%%"

          ## @param snmp_version - integer - optional - default: 2
          ## If you are using SNMP v1 set snmp_version to 1 (required)
          ## If you are using SNMP v3 set snmp_version to 3 (required)
          #
          snmp_version: "%%extra_version%%"

          ## @param timeout - integer - optional - default: 5
          ## Amount of second before timing out.
          #
          timeout: "%%extra_timeout%%"

          ## @param retries - integer - optional - default: 5
          ## Amount of retries before failure.
          #
          retries: "%%extra_retries%%"

          ## @param community_string - string - optional
          ## Only useful for SNMP v1 & v2.
          #
          community_string: "%%extra_community%%"

          ## @param user - string - optional
          ## USERNAME to connect to your SNMP devices.
          #
          user: "%%extra_user%%"

          ## @param authKey - string - optional
          ## Authentication key to use with your Authentication type.
          #
          authKey: "%%extra_auth_key%%"

          ## @param authProtocol - string - optional
          ## Authentication type to use when connecting to your SNMP devices.
          ## It can be one of: MD5, SHA, SHA224, SHA256, SHA384, SHA512.
          ## Default to MD5 when `authKey` is specified.
          #
          authProtocol: "%%extra_auth_protocol%%"

          ## @param privKey - string - optional
          ## Privacy type key to use with your Privacy type.
          #
          privKey: "%%extra_priv_key%%"

          ## @param privProtocol - string - optional
          ## Privacy type to use when connecting to your SNMP devices.
          ## It can be one of: DES, 3DES, AES, AES192, AES256, AES192C, AES256C.
          ## Default to DES when `privKey` is specified.
          #
          privProtocol: "%%extra_priv_protocol%%"

          ## @param context_engine_id - string - optional
          ## ID of your context engine; typically unneeded.
          ## (optional SNMP v3-only parameter)
          #
          context_engine_id: "%%extra_context_engine_id%%"

          ## @param context_name - string - optional
          ## Name of your context (optional SNMP v3-only parameter).
          #
          context_name: "%%extra_context_name%%"

          ## @param loader - string - optional
          ## Check loader to use. Available loaders for snmp:
          ## - core: will use corecheck SNMP integration
          ## - python: will use python SNMP integration
          #
          loader: "%%extra_loader%%"

          ## @param tags - list of key:value element - optional
          ## List of tags to attach to every metric, event and service check emitted by this integration.
          ##
          ## Learn more about tagging: https://docs.datadoghq.com/tagging/
          #
          tags:
            # The autodiscovery subnet the device is part of.
            # Used by Agent autodiscovery to pass subnet name.
            - "autodiscovery_subnet:%%extra_autodiscovery_subnet%%"

          ## @param extra_tags - string - optional
          ## Comma separated tags to attach to every metric, event and service check emitted by this integration.
          ## Example:
          ##  extra_tags: "tag1:val1,tag2:val2"
          #
          extra_tags: "%%extra_tags%%"


  ## @param datadog-cluster.yaml - object - optional
  ## Specify custom contents for the datadog cluster agent config (datadog-cluster.yaml).
  #
  datadog_cluster_yaml:

   node_expiration_timeout: 300

   listeners:
     - name: snmp

   # See here for all `snmp_listener` configs: https://github.com/DataDog/datadog-agent/blob/master/pkg/config/config_template.yaml
   snmp_listener:
     workers: 1000
     discovery_interval: 3600
     configs:
       - network: 172.26.0.0/29
         port: 1161
         community: f5-big-ip
         context_name: 'context1'
         tags:
          - mytag1:val1
          - bar:baz
         loader: core
```

</details>

Make sure to change Agent image to RC image tag. For ClusterAgent, we can keep the custom image I used since there is no DCA release this time.

```
  image:
    repository: datadog/agent-dev  <= to change
    tag: alex-snmp-add-tags-to-snmp-listener-config-extra-tags-py3  <= to change
```

b/

```
helm install datadog-monitoring-core --set datadog.apiKey=XXXXXXXXXXX -f cluster-agent-values_snmp_listener.yaml  --create-namespace -n snmp-core datadog/datadog
```

c/ bash into the agent container
Example:

```
docker exec -it k8s_agent_datadog-monitoring-core-xmpdc_snmp-core_XXXXXX bash
agent check snmp  # check tags from `snmp_listener.config.tags` are here
```

### CODE REVIEW

File: pkg/autodiscovery/listeners/snmp.go [hunk 1]
```diff
@@ -415,6 +417,19 @@ func (s *SNMPService) GetExtraConfig(key []byte) ([]byte, error) {
 		return []byte(s.config.Network), nil
 	case "loader":
 		return []byte(s.config.Loader), nil
+	case "tags":
+		return []byte(convertToCommaSepTags(s.config.Tags)), nil
 	}
 	return []byte{}, ErrNotSupported
 }
+
+func convertToCommaSepTags(tags []string) string {
```
Comment (pducolin) [inline pkg/autodiscovery/listeners/snmp.go:426]: I'm not sure I understand this.

`tags` can be 
1) an array of `key:value` 
2) a string of comma-separated elements, such as `key1:value1,key2:value2`

This function is fine for case `1`, but I don't understand how it works for case `2`. Shouldn't we first split `s.config.Tags` on `,` ?

Comment (AlexandreYang) [reply pkg/autodiscovery/listeners/snmp.go:426]: @pducolin The config tags are like this: 

https://github.com/DataDog/datadog-agent/pull/7680/files#diff-54638e4af140aed843ac96c434105b678d4a402bea72e319fb05e0781f6ed459R203-R206

Then we replace `,` by `_` if needed.

Then the tags are joined with comma `,` and used in template.

An example:

If the `snmp_listener` config is like:

```
snmp_listener:
  configs:
    - network: 192.168.1.16/29
      version: 2
      port: 1161
      community: cisco_icm
      tags:
      - "tag1:val1"
      - "tag2:val2"
```

The Config.tags will be like `[]string{"tag1:val1", "tag2:val2"}` 

Then joined with `,` to be "tag1:val1,tag2:val2"

Then using the template, the result will be like:

```
    ip_address: 192.168.1.17
    ...
    extra_tags: "tag1:val1,tag2:val2"
```

File: pkg/autodiscovery/listeners/snmp.go [hunk 2]
```diff
@@ -415,6 +417,19 @@ func (s *SNMPService) GetExtraConfig(key []byte) ([]byte, error) {
 		return []byte(s.config.Network), nil
 	case "loader":
 		return []byte(s.config.Loader), nil
+	case "tags":
```
Comment (AlexandreYang) [inline pkg/autodiscovery/listeners/snmp.go:420]: note: this correspond to the second "tags" in `extra_tags: "%%extra_tags%%"`

**Review by gmarabout — APPROVED** [review]
ok

**Review by pducolin — APPROVED** [review]

File: pkg/autodiscovery/listeners/snmp.go (hunk 1 above)
Comment (pducolin) [reply pkg/autodiscovery/listeners/snmp.go:426]: I see, the `snmp_listener` is passing the extra_tags configuration to the core_check as a yaml - thought it was passing an object

**Pull request merged**